service/grails-app/services/org/olf/rs/PatronNoticeService.groovy

```diff
@@ -0,0 +1,132 @@
+    // be consumed under the aegis of a timer request
+    log.debug("Configuring event consumer for patron notice service")
+    Properties props = new Properties()
+    try {
+      grailsApplication.config.events.consumer.toProperties().each { final String key, final String value ->
+        // Directly access each entry to cause lookup from env
+        String prop = grailsApplication.config.getProperty("events.consumer.${key}")
+        log.debug("Configuring event consumer for patron notice service :: key:${key} value:${value} prop:${prop}");
+        props.setProperty(key, prop)
+      }
+      log.debug("Configure consumer ${props}")
+    }
+    catch ( Exception e ) {
+      log.error("Problem assembling props for consume",e);
+    }
+
+    consumer = new KafkaConsumer(props)
+    log.debug("PatronNoticeService::init() returning");
+  }
+
+  @javax.annotation.PreDestroy
+  private void cleanUp() throws Exception {
+    log.info("PatronNoticeService::cleanUp");
+    consumer.wakeup();
+  }
+
+  public void triggerNotices(PatronRequest pr, String trigger) {
+    if (!(pr?.patronEmail instanceof String)) {
+      log.error("Notice triggered for request ${pr.hrid} with no email for patron");
+      return;
+    }
+    log.debug("triggerNotices(${pr.patronEmail}, ${trigger})")
+    String tenant = Tenants.currentId()
+    eventPublicationService.publishAsJSON(
+      "${tenant}_PatronNoticeEvents",
+      null,
+      [
+        event:'TriggerNotices_ind',
+        tenant: tenant,
+        payload:[
+          trigger: trigger,
+          email: pr.patronEmail,
+          values: [
+            user: [
+              id: pr?.patronReference,
+              firstName: pr?.patronGivenName,
+              lastName: pr?.patronSurname,
+            ],
+            request: [
+              id: pr.hrid,
+              // pickupLocation: pr?.pickupLocation,
+              // neededBy: pr?.neededBy
+            ],
+            item: [
+              title: pr?.title
+            ]
+          ]
+        ]
+      ]
+    );
+  }
+
+  public void processQueue() {
+    log.debug("Processing patron notice queue")
+    def tenant_list = eventConsumerService.getTenantList()
+    if ( ( tenant_list == null ) || ( tenant_list.size() == 0 ) ) return
+    def topics = tenant_list.collect { "${it}_mod_rs_PatronNoticeEvents".toString() }
+    consumer.subscribe(topics)
+    def consumerRecords = consumer.poll(0)
+    consumerRecords.each{ record ->
+      try {
+        log.debug("KAFKA_EVENT:: topic: ${record.topic()} Key: ${record.key()}, Partition:${record.partition()}, Offset: ${record.offset()}, Value: ${record.value()}");
+        def jsonSlurper = new JsonSlurper()
+        def data = jsonSlurper.parseText(record.value())
+        Tenants.withId(data.tenant) {
+          def notices = NoticePolicyNotice.findAll { noticePolicy.active == true && trigger.value == data.payload.trigger }
```

Comment:
There really should just be one of these. For now. But perhaps once we have the rules engine allowing various notice policies to take effect one might legitimately want certain parties to receive multiple (different) notices in response to the same trigger? So I don't explicitly check for that and error with the idea that I'll enforce (for now) a single active policy elsewhere... But I could do easily enough if someone thinks we should?